fix(coderd/database): consider tag sets when calculating queue position

evgeniy-scherbina opened the pull request:
Relates to https://github.com/coder/coder/issues/15843

## PR Contents

- Reimplementation of the `GetProvisionerJobsByIDsWithQueuePosition` SQL query to **take into account** provisioner job tags and provisioner daemon tags.
- Unit tests covering different **tag sets**, **job statuses**, and **job ordering** scenarios.

## Notes

- The original row order is preserved by introducing the `ordinality` field.
- Unnecessary rows are filtered as early as possible to ensure that expensive joins operate on a smaller dataset.
- A "fake" join with `provisioner_jobs` is added at the end to ensure `sqlc.embed` compiles successfully.
- **Backward compatibility is preserved**—only the SQL query has been updated, while the Go code remains unchanged.

evgeniy-scherbina committed:
fix(coderd/database): remove linux build tags from db package

evgeniy-scherbina committed:
fix(coderd/database): consider tag sets when calculating queue position

evgeniy-scherbina committed:
test(coderd/database): skip tests when PostgreSQL is unavailable

evgeniy-scherbina committed:
test(coderd/database): create default provisioner in tests

Emyrk commented on coderd/database/querier.go:
```diff
@@ -214,7 +214,9 @@ type sqlcQuerier interface {
 	GetProvisionerJobByID(ctx context.Context, id uuid.UUID) (ProvisionerJob, error)
 	GetProvisionerJobTimingsByJobID(ctx context.Context, jobID uuid.UUID) ([]ProvisionerJobTiming, error)
 	GetProvisionerJobsByIDs(ctx context.Context, ids []uuid.UUID) ([]ProvisionerJob, error)
+	// Step 5: Final SELECT with INNER JOIN provisioner_jobs
 	GetProvisionerJobsByIDsWithQueuePosition(ctx context.Context, ids []uuid.UUID) ([]GetProvisionerJobsByIDsWithQueuePositionRow, error)
+	// Preserve original ID order from upstream
 	GetProvisionerJobsByOrganizationAndStatusWithQueuePositionAndProvisioner(ctx context.Context, arg GetProvisionerJobsByOrganizationAndStatusWithQueuePositionAndProvisionerParams) ([]GetProvisionerJobsByOrganizationAndStatusWithQueuePositionAndProvisionerRow, error)
```
SQLc includes all leftmost indented comments in sql. So you have to add a space before them to omit the comments here :cry:

Emyrk commented on coderd/database/querier_test.go:
```diff
@@ -2164,11 +2171,125 @@ func TestGetProvisionerJobsByIDsWithQueuePosition(t *testing.T) {
+				{"a": "1", "b": "2"},
+				{"a": "1"},
+			},
+			queueSizes:     []int64{2, 2, 0},
+			queuePositions: []int64{1, 1, 0},
```
This is interesting. If there are 0 provisioners for a given job, we return `0/0`?

evgeniy-scherbina replied:
yes, my decision was based on previous behaviour: https://github.com/coder/coder/blob/main/coderd/database/querier_test.go#L2267-L2277
previously if the job isn't pending (for ex. completed), we return `0 / global queue size`, for ex. `0 / 2`

But it's a good question, now we have two different use-cases when we return `0/0`
- job isn't pending -> `0/0`
- job is pending, but doesn't have a provisioner -> `0/0`

Probably we need to differentiate them, maybe return the `status` along with `queue_pos` & `queue_size`

evgeniy-scherbina replied:
We decided to keep `0/0` for now, and change it later if it will be necessary.

Emyrk commented on coderd/database/querier_test.go:
```diff
@@ -2164,11 +2171,125 @@ func TestGetProvisionerJobsByIDsWithQueuePosition(t *testing.T) {
+	now := dbtime.Now()
+	ctx := testutil.Context(t, testutil.WaitShort)
+
+	testCases := []struct {
+		name           string
+		jobTags        []database.StringMap
+		daemonTags     []database.StringMap
+		queueSizes     []int64
+		queuePositions []int64
+	}{
```
I really like this testCase structure :heart:

evgeniy-scherbina replied:
I extended it with:

```go
// GetProvisionerJobsByIDsWithQueuePosition takes jobIDs as a parameter.
// If skipJobIDs is empty, all jobs are passed to the function; otherwise, the specified jobs are skipped.
// NOTE: Skipping job IDs means they will be excluded from the result,
// but this should not affect the queue position or queue size of other jobs.
skipJobIDs map[int]struct{}
```
to cover additional test cases.

Emyrk commented on coderd/database/querier_test.go:
```diff
@@ -1258,6 +1256,15 @@ func TestQueuePosition(t *testing.T) {
+	dbgen.ProvisionerDaemon(t, db, database.ProvisionerDaemon{
+		Name:         "default_provisioner",
+		Provisioners: []database.ProvisionerType{database.ProvisionerTypeEcho},
+		// Ensure the `tags` field is NOT NULL for the default provisioner;
+		// otherwise, it won't be able to pick up any jobs.
+		Tags: database.StringMap{"a": "1"},
```
Does `null` tags break things?

evgeniy-scherbina replied:
`<@` - Checks if the left operand is contained within the right operand.

```SQL
SELECT NULL::jsonb <@ NULL::jsonb; -> returns NULL
SELECT NULL::jsonb <@ '{}'::jsonb; -> returns NULL
SELECT '{}'::jsonb <@ '{}'::jsonb; -> return true
```

so `provisioner_tagset_contains(pd.tags, pj.tags)` returns NULL if either of arguments is NULL

My assumption was that in production we don't have NULL tags, so we don't have to worry about it?
Also I think I can use `{}` instead of `{"a": "1"}`

evgeniy-scherbina replied:
I see that we already use this approach in other places: https://github.com/coder/coder/blob/main/coderd/database/querier_test.go#L1250

evgeniy-scherbina replied:
slightly improved it: https://github.com/coder/coder/pull/16685/commits/c1f421dc990c096d4f586e5c5082e4666db0bc32

Emyrk commented on coderd/database/querier_test.go:
```diff
@@ -2164,11 +2171,125 @@ func TestGetProvisionerJobsByIDsWithQueuePosition(t *testing.T) {
+		},
+	}
+
+	for _, tc := range testCases {
+		db, _ := dbtestutil.NewDB(t)
+
```
Can we throw this into subtests?

```suggestion
	for _, tc := range testCases {
          tc := tc
          t.Run(tc.name, func(t *testing.T) {
                t.Parallel()
		db, _ := dbtestutil.NewDB(t)

```

evgeniy-scherbina replied:
I also thought about that. Done.

Emyrk commented on coderd/database/dbtestutil/postgres_test.go:
```diff
@@ -44,6 +48,9 @@ func TestOpen_InvalidDBFrom(t *testing.T) {
 
 func TestOpen_ValidDBFrom(t *testing.T) {
 	t.Parallel()
+	if !dbtestutil.WillUsePostgres() {
+		t.Skip("this test requires postgres")
+	}
```
Flipping this generates a bit less test output in the logs
```suggestion
	if !dbtestutil.WillUsePostgres() {
		t.Skip("this test requires postgres")
	}
	t.Parallel()
```

evgeniy-scherbina replied:
I changed it in newly added tests, but I noticed that across database package we use:
```go
t.Parallel()
if !dbtestutil.WillUsePostgres() {
    t.Skip("this test requires postgres")
}
```

evgeniy-scherbina replied:
@Emyrk 
Turns out it fails the linter: https://github.com/coder/coder/actions/runs/13570489300/job/37933949541

`Error: coderd/database/querier_test.go:2169:6: TestGetProvisionerJobsByIDsWithQueuePosition should call t.Parallel on the top level as well as its subtests (tparallel)`

Emyrk replied:
TIL :+1:

Emyrk commented on coderd/database/queries/provisionerjobs.sql:
```diff
@@ -50,42 +50,67 @@ WHERE
+	GROUP BY
+		fpj.id, fpj.created_at, fpj.ordinality -- Include `ordinality` in GROUP BY
 )
+-- Step 5: Final SELECT with INNER JOIN provisioner_jobs
```
Formatting nit. If you have this leftmost indented, then it appears in the sqlc interface as a leading comment. I have another comment mentioning this.

```suggestion
 -- Step 5: Final SELECT with INNER JOIN provisioner_jobs
```

evgeniy-scherbina replied:
Fixed it, I tried to find a way how to have better control over, but looks like indentation is the only way.

Emyrk replied:
Unfortunately :cry:

Emyrk commented on coderd/database/queries/provisionerjobs.sql:
```diff
@@ -50,42 +50,67 @@ WHERE
+	-- Step 1: Filter provisioner_jobs and assign an order from upstream system
+	SELECT
+		*,
+		ROW_NUMBER() OVER () AS ordinality -- Track original order
```
Is `ROW_NUMBER()` always the correct rank index? Should they be ordered by `created_at`?

Emyrk commented on coderd/database/queries/provisionerjobs.sql:
```diff
@@ -50,42 +50,67 @@ WHERE
+		fpj.ordinality -- Preserve original order
+	FROM
+		filtered_provisioner_jobs fpj -- Use the pre-filtered dataset instead of full provisioner_jobs
+			LEFT JOIN ranked_jobs rj
+					ON fpj.id = rj.id -- Ensure we only keep jobs that have a ranking
```
> Ensure we only keep jobs that have a ranking

According to the comment, would this be an `INNER JOIN`? 


----


Saying that, I think the `LEFT JOIN` is correct here. Non-ranked jobs are jobs that do not match any existing provisioner. This can, and does happen. We want to return these jobs. So I would agree with `LEFT JOIN`, just need to update the comment :+1:

evgeniy-scherbina replied:
Agree, it was confusing. Done.

Emyrk commented on coderd/database/queries/provisionerjobs.sql:
```diff
@@ -50,42 +50,67 @@ WHERE
+	SELECT
+		fpj.id,
+		fpj.created_at,
+		COALESCE(MIN(rj.queue_position), 0) :: BIGINT AS queue_position, -- Best queue position across provisioners
+		COALESCE(MAX(rj.queue_size), 0) :: BIGINT AS queue_size, -- Max queue size across provisioners
```
I understand this `0`. Should we instead use `-1` to indicate no valid `queue_position` or `queue_size`? 

0 could be misinterpreted as "complete" imo. `-1` is more obviously "not in a queue".

Emyrk reviewed (commented):
This is looking good, the SQL query is compelx, but well structured.

Can we ensure we have unit tests for the following cases:
- N jobs (1 job with 0 tags) & 0 provisioners exist
- N jobs (1 with 0 tags) & N provisioners
  - 1 job not matching **any** provisioner
- 0 jobs & 0 provisioners
- N jobs with at least 1 being completed (not pending)

Just want to cover all cases

evgeniy-scherbina committed:
fix(coderd/database): correctly calculate pending and ranked jobs

evgeniy-scherbina committed:
test(coderd/database): improve test coverage

evgeniy-scherbina committed:
refactor(coderd/database): use empty tag sets

evgeniy-scherbina committed:
test(coderd/database): improve test coverage

evgeniy-scherbina commented:
@Emyrk I added test cases according to your recommendations: https://github.com/coder/coder/pull/16685/commits/e7693eeca6733ca404f0649870ba096926fe45f8

Also in previous version I made a mistake, I filtered out provisioner jobs by passed `@ids` too early, meaning if `jobID` is not passed to `GetProvisionerJobsByIDsWithQueuePosition` it won't be considered during `queue_pos` & `queue_size` calculation - so I fixed it and added test cases to cover it, they use this approach:

```go
// GetProvisionerJobsByIDsWithQueuePosition takes jobIDs as a parameter.
// If skipJobIDs is empty, all jobs are passed to the function; otherwise, the specified jobs are skipped.
// NOTE: Skipping job IDs means they will be excluded from the result,
// but this should not affect the queue position or queue size of other jobs.
skipJobIDs map[int]struct{}
```

evgeniy-scherbina committed:
test(coderd/database): wrap test cases in subtests

evgeniy-scherbina committed:
refactor(coderd/database): minor refactoring in tests

evgeniy-scherbina committed:
fix(coderd/database): linter

evgeniy-scherbina committed:
refactor(coderd/database): use job_status generated column

evgeniy-scherbina committed:
perf: add index on provisioner_jobs table for status field

evgeniy-scherbina committed:
Merge remote-tracking branch 'origin/main' into 15843-queue-position

evgeniy-scherbina committed:
fix(coderd/database): increment migration number

evgeniy-scherbina committed:
fix(coderd/database): update dump.sql

evgeniy-scherbina committed:
perf: optimize get-queue-position sql query

Emyrk reviewed (approved):
LGTM! :+1:

Emyrk reviewed (dismissed):
Sorry forgot about dbmem. :cry: 

SQL and tests look good :+1:

evgeniy-scherbina committed:
chore: update dbmem

evgeniy-scherbina committed:
Merge remote-tracking branch 'origin/main' into 15843-queue-position

evgeniy-scherbina committed:
fix(coderd/database): increment migration number

johnstcn commented on coderd/database/migrations/000298_provisioner_jobs_status_idx.up.sql:
```diff
@@ -0,0 +1 @@
+CREATE INDEX idx_provisioner_jobs_status ON provisioner_jobs USING btree (job_status);
```
Review: it took me a minute to figure out if this is supported, but apparently it is! The [docs on generated columns](https://www.postgresql.org/docs/16/ddl-generated-columns.html) specify that only stored generated columns are supported, and these are always written to disk. TIL!

evgeniy-scherbina replied:
@johnstcn 
yes, that's what I figured as well
Also I thought about using `Hash` Index instead of `BTree`, because so far we're only using `=` or `!=` operations, but wasn't sure, maybe we'll need other operations in the future.

johnstcn replied:
TBH this is probably fine, almost all of our other indexes are using BTree. If you were curious, I suppose you could validate with some EXPLAINs.

johnstcn commented on coderd/database/querier_test.go:
```diff
@@ -2159,6 +2168,307 @@ func TestExpectOne(t *testing.T) {
+			name:           "test-case-12",
+			jobTags:        []database.StringMap{},
+			daemonTags:     []database.StringMap{},
+			queueSizes:     nil, // TODO(yevhenii): should it be empty array instead?
```
In general I prefer to return an empty slice for something that is always expected to be present. However, as these responses end up being converted to Typescript objects, we kind of always have to specify arrays as being nullable due to Go's type system. @Emyrk may have some more thoughts here.

Emyrk replied:
This is for a test case. I also generally prefer an empty slice in this case for the equality assertion. But I have no strong opinion in this test case.

---

For the typescript side of things, (off the top of my head an quick glancing) slices are never `null`. So a `nil` slice --> `[]` on the typescript side of things.

https://github.com/coder/coder/blob/6f4da846464915fa8899fb93307edbd6bbc73cf9/site/src/api/typesGenerated.ts#L185-L185

johnstcn commented on coderd/database/migrations/000298_provisioner_jobs_status_idx.up.sql:
Obligatory reminder to check migration number before merge!

johnstcn reviewed (approved):
👍 Changes look good to me! Nice tests!
